wgsl/index.bs

```diff
@@ -1366,86 +1375,360 @@ Therefore, a data value must not be placed in the padding at the end of a struct
+<div class='example wgsl' heading='Pointer from a variable'>
+  <xmp highlight='rust'>
+    // Declare a variable in the private storage class, for storing an f32 value.
+    var<private> x: f32;
+
+    fn f() -> void {
+        // Declare a variable in the function storage class, for storing an i32 value.
+        var y: i32;
+
+        // The name 'x' resolves to the module-scope variable 'x',
+        // and has reference type ref<private,f32>.
+        // Applying the unary '&' operator converts the reference to a pointer.
+        const x_ptr: ptr<private,f32> = &x;
+
+        // The name 'y' resolves to the function-scope variable 'y',
+        // and has reference type ref<private,i32>.
+        // Applying the unary '&' operator converts the reference to a pointer.
+        const y_ptr: ptr<function,i32> = &y;
+
+        // A new variable, distinct from the variable declared at module scope.
+        var x: u32;
+
+        // Here, the name 'x' resolves to the function-scope variable 'x' declared in
+        // the previous statement, and has type ref<function,u32>.
+        // Applying the unary '&' operator converts the reference to a pointer.
+        const inner_x_ptr: ptr<function,u32> = &x;
+    }
+  </xmp>
+</div>
 
----
 
-A pointer may appear in exactly the following contexts
+### Comparison with references and pointers in other languages ### {#pointers-other-languages}
 
-<table class='data'>
-  <tr><td>Indexing<td>
-A subaccessing evaluation
-* E.g. `a[12]`
-    * If `a` is a pointer to an array, this evaluates to *a.Subaccess(12)*
-
-* E.g. `s.foo`
-    * If `s` is a pointer to a structure of type *S*, `k` is the index of the `foo` element of *S*, this evaluates to *s.Subaccess(k)*
-
-  <tr><td>Assigning (L-Value)<td>
-On the left hand side of an assignment operation, and the right hand side
-matches the pointee type of the pointer.
-* E.g. `v = 12;` assuming prior declaration `var v : i32`
-
-  <tr><td>Copying<td>
-On the right hand side of a const-declaration, and the type of the
-const-declaration matches the pointer type.
-* E.g. `const v2 : ptr<private,i32> = v;`  assuming prior declaration
-        `var<private> v:i32`
-
-  <tr><td>Parameter<td>
-Used in a function call, where the function’s parameter type matches the
-pointer type.
-
-  <tr><td>Reading (R-Value)<td>
-Any other context.  Evaluates to *P.Read()*, yielding a value of *P*’s pointee
-type.
-</table>
+This section is informative, not normative.
+
+References and pointers in [SHORTNAME] are more restricted than in other languages.
+In particular:
+
+* In [SHORTNAME] a reference can't directly be declared as an alias to another reference or variable,
+    either as a variable or as a formal parameter.
+* In [SHORTNAME] pointers and references are not [=storable=].
+    That is, the content of a [SHORTNAME] variable may not contain a pointer or a reference.
+* In [SHORTNAME] a function must not return a pointer or reference.
```

Comment:
nit: Maybe we should explicitly state (or just give an example) that return types, unlike `var` are true-declared-types and not references, and `return v;` where `v` is of type `ref`, will perform an automatic dereference?

Reply:
In parallel we landed:
- #1568 which renames "value types" to "plain types", and
- #1570 which limits return types on functions to plain types.

So what's left is an example showing that if 'return foo;' is used and foo is a reference, that we apply the Load Rule and return the value read via foo.